nvflare/app_common/workflows/wf_controller.py

```diff
@@ -0,0 +1,296 @@
+    CMD_RELAY,
+    CMD_SEND,
+    CMD_STOP,
+    DATA,
+    MIN_RESPONSES,
+    PAYLOAD,
+    RESULT,
+    SITE_NAMES,
+    STATUS,
+    TARGET_SITES,
+)
+from nvflare.app_common.workflows.wf_comm.wf_queue import WFQueue
+from nvflare.app_common.workflows.wf_comm.wf_spec import WF
+from nvflare.fuel.utils import class_utils
+from nvflare.security.logging import secure_format_traceback
+
+
+class WFController(ErrorHandlingController):
+    def __init__(
+        self,
+        task_name: str,
+        wf_class_path: str,
+        wf_args: Dict,
+        task_timeout: int = 0,
+        comm_msg_pull_interval: float = 0.2,
+    ):
+        super().__init__()
+
+        self.clients = None
+        self.task_timeout = task_timeout
+        self.task_name = task_name
+        self.comm_msg_pull_interval = comm_msg_pull_interval
+        self.wf_class_path = wf_class_path
+        self.wf_args = wf_args
+        self.wf_queue: WFQueue = WFQueue(ctrl_queue=Queue(), result_queue=Queue())
+        self.wf: WF = class_utils.instantiate_class(self.wf_class_path, self.wf_args)
+        self._thread_pool_executor = ThreadPoolExecutor(max_workers=2, thread_name_prefix=self.__class__.__name__)
+
+        self.engine = None
+        self.fl_ctx = None
+
+    def start_controller(self, fl_ctx: FLContext):
+        self.fl_ctx = fl_ctx
+        self.log_info(fl_ctx, "Initializing controller workflow.")
+        self.engine = self.fl_ctx.get_engine()
+        self.clients = self.engine.get_clients()
+
+        self.setup_wf_queue()
+
+        self.log_info(fl_ctx, "workflow controller started")
+
+    def setup_wf_queue(self):
+        wf_comm_api = self.find_wf_comm_in_wf()
+        wf_comm_api.set_queue(self.wf_queue)
+        wf_comm_api.set_result_pull_interval(self.comm_msg_pull_interval)
+        wf_comm_api.meta.update({SITE_NAMES: self.get_site_names()})
+
+        print(f"\n \n  {wf_comm_api.meta=}")
+
+    def find_wf_comm_in_wf(self):
+        attr_objs = [getattr(self.wf, attr_name, None) for attr_name in dir(self.wf)]
+        wf_comm_attrs = [attr for attr in attr_objs if isinstance(attr, WFCommAPI)]
+        if wf_comm_attrs:
+            return wf_comm_attrs[0]
+        else:
+            raise RuntimeError(f"missing required attribute with type of 'WFCommAPI' in {self.wf.__class__.__name__}")
+
+    def control_flow(self, abort_signal: Signal, fl_ctx: FLContext):
+        try:
+            future = self._thread_pool_executor.submit(self.ctrl_msg_loop, fl_ctx=fl_ctx, abort_signal=abort_signal)
+            self.wf.run()
+            self.stop_msg_queue("job completed", fl_ctx)
+            future.result()
+        except Exception as e:
+            error_msg = secure_format_traceback()
+            self.log_error(fl_ctx, error_msg)
+            self.system_panic(error_msg, fl_ctx=fl_ctx)
+        finally:
+            wait_time = self.comm_msg_pull_interval + 0.05
+            self.stop_msg_queue("job finished", fl_ctx, wait_time)
```

Comment:
stop_msg_queue could be called twice (a normal 'job completed' and here 'job finished').  The logic of this function supports re-entrace, but it could confuse users.